### PR TITLE
Api keys and tokens rule

### PR DESCRIPTION
**Pull request by Njeriiii**
What does this implement?
This class is a custom inspection tool that checks for hardcoded API keys and tokens in the code.
It extends the LocalInspectionTool class and overrides the buildVisitor method to create a visitor for the inspection.
The visitor checks for the use of specific methods that are known to be used for API keys and tokens.
These are AzureKeyCredential and AccessToken.
 

Has this been tested?
---------------------------
- [x] Tested

Any relevant logs, screenshots, error output, etc.?
-------------------------------------
image

image

image

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json [hunk 1]
```diff
@@ -1,6 +1,12 @@
 {
-  "StorageUploadWithoutLengthCheck": {
-    "methodsToCheck": [
+  "HardcodedAPIKeysAndTokensCheck": {
+    "clients_to_check": ["AzureKeyCredential", "AccessToken"],
+    "anti_pattern_message": "Use of API keys or tokens in code is not recommended. We recommend Secretless authentication through DefaultAzureCredential.",
```
Comment (g2vinay) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: Better to phrase this along these lines 
DefaultAzureCredential is recommended for authentication if the service client supports Token Credential (Entra ID Authentication). if not, then use Azure Key Credential for API key based authentication.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json [hunk 2]
```diff
@@ -1,6 +1,12 @@
...
+    "solution": "We recommend Secretless authentication through DefaultAzureCredential."
```
Comment (g2vinay) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: same as above.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json (hunk 1 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: done:)

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json (hunk 2 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: done:)

File: PluginsAndFeatures/azure-toolkit-for-intellij/.idea/codeStyles/Project.xml
Comment (Njeriiii) [inline PluginsAndFeatures/azure-toolkit-for-intellij/.idea/codeStyles/Project.xml:1]: not sure why these .idea files are committed. they're ignored in the .gitignore and also don't show up in a specific commit in this PR

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/readme.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Project: Java Code Quality Analyzer
+
+## Project Description
+The Java Code Quality Analyzer, is a plugin designed to improve the quality of Java code. It provides an interactive tool window that offers real-time code suggestions. The following are proposed essential features, including rule set integration, telemetry connectivity, and Azure Toolkit integration.
```
Comment (billwert) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/readme.md]:
```suggestion
The Java Code Quality Analyzer is a plugin designed to improve the quality of Java code. It provides an interactive tool window that offers real-time code suggestions. The following are proposed essential features, including rule set integration, telemetry connectivity, and Azure Toolkit integration.
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/readme.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Key Features
+
+- **Rule Set Integration**: This feature enables users to import and customize rule sets, tailoring the plugin to their specific needs.
+- **Telemetry Integration**: This feature connects the plugin to the backend with Application Insights, allowing for efficient data transmission.
+- **Azure Toolkit Integration**: This feature allows the plugin to integrate with the IntelliJ Azure toolkit, providing extended functionality.
```
Comment (billwert) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/readme.md]:
```suggestion
- **Azure Toolkit for IntelliJ Integration**: This feature allows the plugin to integrate with the Azure Toolkit for IntelliJ, providing extended functionality.
```

Correcting to the real product name

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json [hunk 5]
```diff
@@ -10,5 +10,11 @@
   "ServiceBusReceiverAsyncClientCheck": {
     "client_name": "ServiceBusReceiverAsyncClient",
     "antipattern_message": "Use of ServiceBusReceiverAsyncClient detected. Use ServiceBusProcessorClient instead."
+  },
+
+  "HardcodedAPIKeysAndTokensCheck": {
+    "clients_to_check": ["AzureKeyCredential", "AccessToken"],
```
Comment (billwert) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: Consider adding `KeyCredential`, `AzureNamedKeyCredential`, `AzureSasCredential`, `AzureNamedKey`, `ClientSecretCredential`, `UsernamePasswordCredential`, and `BasicAuthenticationCredential`. All of these are similar in that they might encourage hard coding a secret in some form.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
+package com.microsoft.azure.toolkit.intellij.azure.sdk.buildtool;
+
+import com.intellij.codeInspection.LocalInspectionTool;
+import com.intellij.codeInspection.ProblemsHolder;
+import com.intellij.psi.JavaElementVisitor;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiElementVisitor;
+import com.intellij.psi.PsiNewExpression;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONObject;
```
Comment (billwert) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java:10]: Can we take a dependency on `azure-json` here and use that instead?

**Review by billwert — APPROVED** [review]
Left a handful of suggestions, none of which is critical. LGTM.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        private static Map<String, Object> loadRuleConfigurations() {
+
+            // Define constants for string literals
+            final String RULE_KEY = "HardcodedAPIKeysAndTokensCheck";
+            final String ANTI_PATTERN_MESSAGE_KEY = "anti_pattern_message";
+            final String CLIENTS_TO_CHECK_KEY = "clients_to_check";
+            final String RULE_CONFIGURATION = "META-INF/ruleConfigs.json";
+
+            // Load the rule configurations from the JSON file
+            try {
```
Comment (g2vinay) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java]: This section can be updated to ensure input stream is properly closed as well.

```java
try (InputStream inputStream = HardcodedAPIKeysAndTokensCheck.class.getClassLoader().getResourceAsStream(RULE_CONFIGURATION)) {
    if (inputStream == null) {
        throw new FileNotFoundException("Configuration file not found");
    }
    // Read the JSON object from the input stream
    try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream))) {
        final JSONObject jsonObject = new JSONObject(bufferedReader.lines().collect(Collectors.joining()));
        final JSONObject apiKeysAndTokensObject = jsonObject.getJSONObject(RULE_KEY);
        final String antiPatternMessage = apiKeysAndTokensObject.getString(ANTI_PATTERN_MESSAGE_KEY);
        final List<String> clientsToCheck = apiKeysAndTokensObject.getJSONArray(CLIENTS_TO_CHECK_KEY).toList().stream().map(Object::toString).collect(Collectors.toList());
        return Map.of(ANTI_PATTERN_MESSAGE_KEY, antiPatternMessage, CLIENTS_TO_CHECK_KEY, clientsToCheck);
    }
} catch (Exception e) {
    e.printStackTrace();
    return Collections.emptyMap();
}
```

**Review by g2vinay — APPROVED** [review]

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java (hunk 6 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java:10]: I'm working on a centralised json loader for all rules that uses azure-json:)

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java (hunk 7 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java]: Thanks! I also refactored this section in a separate PR so I will be removing the json loading functionality from this file: https://github.com/Njeriiii/azure-tools-for-java/pull/16/files#diff-e739175547b0efaff446f64bf6c8f34d16e4e14f275e75ebd38e47a677b0e1b9R84

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java (hunk 6 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheck.java:10]: this is the PR for the json loader :) https://github.com/Njeriiii/azure-tools-for-java/pull/16

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json (hunk 5 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: done:) these 2 use builders for instantiation so I'm checking for the builders: ClientSecretCredentialBuilder, "UsernamePasswordCredentialBuilder

**Review by billwert — APPROVED** [review]
lgtm!

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json [hunk 8]
```diff
@@ -10,5 +10,11 @@
   "ServiceBusReceiverAsyncClientCheck": {
     "client_name": "ServiceBusReceiverAsyncClient",
     "antipattern_message": "Use of ServiceBusReceiverAsyncClient detected. Use ServiceBusProcessorClient instead."
+  },
+
+  "HardcodedAPIKeysAndTokensCheck": {
+    "clients_to_check": ["AzureKeyCredential", "AccessToken", "KeyCredential", "AzureNamedKeyCredential", "AzureSasCredential", "AzureNamedKey", "ClientSecretCredentialBuilder", "UsernamePasswordCredentialBuilder", "BasicAuthenticationCredential"],
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: Why are we calling these clients_to_check, aren't these just different ways to authenticate?
we should consider renaming this to something more accurate.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+    // Helper method to assert visitor is not null and is an instance of JavaElementVisitor
+    void assertVisitor(PsiElementVisitor mockVisitor) {
+
+        assertNotNull(mockVisitor);
+        assertTrue(mockVisitor instanceof JavaElementVisitor);
+    }
+
+    // Helper method to verify registerProblem is called
+    void verifyRegisterProblem(String CLIENT_TO_CHECK, int NUMBER_OF_INVOCATIONS) {
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java]: do we need these methods package-private

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        mockVisitor = createVisitor();
+    }
+
+
+    /**
+     * Test the HardcodedAPIKeysAndTokensCheck class for hardcoded API keys and tokens.
+     * When a client is authenticated with AzurekeyCredentials and AccessToken, a problem is registered.
+     */
+    @Test
+    public void testHardcodedAPIKeysAndTokensCheck() {
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java]: Use ParameterizedTest to test all the variations of authentications we need to flag.
https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        when(javaCodeReferenceElement.getReferenceName()).thenReturn(CLIENT_TO_CHECK);
+
+        when(javaCodeReferenceElement.getQualifiedName()).thenReturn("com.azure");
+
+        mockVisitor.visitElement(newExpression);
+
+        // Verify registerProblem is called
+        verify(mockHolder, times(NUMBER_OF_INVOCATIONS)).registerProblem(eq(newExpression),
+                Mockito.contains("DefaultAzureCredential is recommended for authentication if the service client supports Token Credential (Entra ID Authentication). " +
+                        "If not, then use Azure Key Credential for API key based authentication."));
```
Comment (samvaity) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java]: This seems to be incorrect. We are flagging APIKey Credential pattern but also suggesting it?

Comment (billwert) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java]: This is a reasonable message; we need to allow for circumstances where the user can't use one of the better options (services that only support keys, etc.)

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java (hunk 9 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java]: oh they should be private actually thanks:)

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java (hunk 10 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java]: done:)

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json (hunk 8 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/resources/META-INF/ruleConfigs.json]: changed to services_to_check -- I also made a new getter for it. i thought about combining with the clients to check getter but I think its better to keep everything separate for now incase there's a situation where I have specific clients to look for in certain services (eg the single operations in loop check in the future will be expanded to storage & text analytics)

**Pull request merged**